### ISSUE TITLE
: vs ; in escape sequences

### ISSUE THREAD

**Issue by dkuettel** [terminalguide/terminalguide#9]
I couldn't find an explanation about this issue. I always used semicolons, like `ESC[38;5;7m` even though in https://terminalguide.namepad.de/attr/ it uses colons and not semicolons. At first I thought that's maybe just your notation, but now I'm not sure. Is both valid, or does it depend on my terminal emulator?

The reason I'm asking is because of, eg, the special "curly underline", it's documented as `4:3`. Now I know my terminal (alacritty) doesnt support that, but I wanted to see what happens. If I use `4:3` then it just does normal underline, if I do `4;3` it does underline with italics, as I think it should, `;` combines multiple codes. is `4:3` a "subcode", or are terminals supposed to recognize the sequence `4;3` and then not treat it as a `4` and a `3`?

**Comment by textshell** [comment]
It's actually a fairly complicated topic.
For colors terminals started with using only ';' based on a (in hindsight incorrect) interpretation of a source standard that was at that time hard(or expensive) to obtain. Later reading of that material combined with the compatibility problems of using ';' motivated a change in understanding that ':' would have been right and it would be best to switch over to that for the color sequences. The digits groups following the ':'s are called sub-parameters in terminal guide.

Terminal guide tries to avoid muddling the current way things are interpreted by too much jargon or history. So this is not really described anywhere. But for example in https://terminalguide.namepad.de/attr/bgdirectcolor/ you can see which terminals support which format (further complicated by the discovery that the original source also had a very under-specified color space parameter that got lost for some time). It's a bit of a mess and for colors switching everything to the ':' format seems like a very long term project or even infeasible.

For sequences like the underlining the lessons from the color sequences had already been learned and no ';' variant have been implemented, because they would only create a mess. Thus for underline ':' is mandatory to use and good terminal implementations should be able to ignore these new sequences if they are not supported (yet).

tl;dr: For rgb colors both ':' and ';' are supported by some terminals while other only support a specific style. For everything else ':' is mandatory and could be understood as "subcode".

**Comment by dkuettel** [comment]
Very interesting, thanks for the detailed reply, very much appreciated.

I usually used `ESC[0;4;38;5;7m` to combine, eg, reset+underline+color, using `;` also to combine codes. I always thought that the codes for SGR were chosen to be "prefix free" so that the interpretation is always clear. I see that with `:` for subcode and `;` for combining there would be no ambiguity. Is it in fact legal to combine multiple SGR into one escape using `;`? Or was I just lucky that it seemed to work so far. Not that here and there I would notice if a color was actually off, so maybe it didnt always work.

**Comment by textshell** [comment]
Yes, combining attributes codes is legal. That's documented in https://terminalguide.namepad.de/seq/csi_sm/ 
Maybe it's too subtle?

The code for modes and attributes are supposed to stand on their own, so a terminal can just ignore unknown codes. The encoding with ';' for rgb colors broke that. So now there are special cases and prefix-free is what we get (or something similar restricted to the 2 or 3 codes that have this problem)

**Comment by dkuettel** [comment]
Ah no you're right, it's very concise and clear there that you can repeat codes. Kudos for collecting all the information into a clean documentation.

I read both your replies again, I hope I didnt miss it, what is not 100% clear to me: Is `;` the sequence operator and `:` the subcode operator? Or is `;` in theory never to be used, and just there because of broken terminals?

It's clear from https://terminalguide.namepad.de/seq/csi_sm/ that I can repeat, but I'm not sure what is the correct way to separate codes.
